test/lib/listeners/ProgressEventListener.test.ts

```diff
@@ -0,0 +1,321 @@
+			const engine2Percents = getDedupedCompletionPercentages(displayEvents.slice(2, displayEvents.length - 1), engine2.getName());
+			// Both engines should proceed in the expected percentage order.
+			expect(engine1Percents).toEqual([0, 50, 100]);
+			expect(engine2Percents).toEqual([0, 50, 100]);
+			// The final event should be the Stop event.
+			const endEvent = displayEvents[displayEvents.length - 1];
+			expect(endEvent).toHaveProperty('type', DisplayEventType.SPINNER_STOP);
+			expect(endEvent.data).toContain('done');
+		});
+
+		it('Properly interleaves progress updates with ticking', async () => {
+			// ==== TEST SETUP ====
+			// Use a plugin with engines that can be configured to wait a certain amount of time between sending their
+			// update events.
+			const timeablePlugin = new TimeableStubEnginePlugin1();
+			await codeAnalyzer.addEnginePlugin(timeablePlugin);
+			const engine: TimeableEngine1 = timeablePlugin.getCreatedEngine('timeableEngine1') as TimeableEngine1;
+			// Tell it to wait for a relatively small amount of time (less than a second, so the test doesn't take forever).
+			engine.setEngineExecutionWaitTime(500);
+			// The specific files we use for our workspace don't matter.
+			const workspace = await codeAnalyzer.createWorkspace(['package.json']);
+			// Select only rules from one engine, since this test is about how one engine interleaves with ticking, not
+			// how multiple engines interleave with each other.
+			const ruleSelection = await codeAnalyzer.selectRules(['timeableEngine1'], {workspace});
+			// Set the spinner's tick time to be unrealistically low, to make sure ticks happen between engine update events.
+			spinner = new EngineRunProgressSpinner(spyDisplay, 200);
+
+			// ==== TESTED BEHAVIOR ====
+			// Start listening, execute the rules, and stop listening.
+			spinner.listen(codeAnalyzer);
+			await codeAnalyzer.run(ruleSelection, {
+				workspace
+			});
+			spinner.stopListening();
+
+			// ==== ASSERTIONS ====
+			const displayEvents = spyDisplay.getDisplayEvents();
+			const startEvent = displayEvents[0];
+			expect(startEvent).toHaveProperty('type', DisplayEventType.SPINNER_START);
+			const percentagesInOrder = getDedupedCompletionPercentages(displayEvents.slice(1, displayEvents.length - 1));
+			expect(percentagesInOrder).toEqual([0, 50, 100]);
+			const endEvent = displayEvents[displayEvents.length - 1];
+			expect(endEvent).toHaveProperty('type', DisplayEventType.SPINNER_STOP);
+			expect(endEvent.data).toContain('done');
+		}, 10000);
+	});
+
+	/**
+	 * It's possible (and in fact quite likely) that a list of display events includes redundant entries. For example, the
+	 * analyzer and an engine could both send a 0% or 100% completion event.
+	 * This helper method removes repeated completion percentages from the display events, to keep the tests flexible.
+	 */
+	function getDedupedCompletionPercentages(displayEvents: DisplayEvent[], engine?: string): number[] {
```

Comment:
The previous version of this method was overzealous. We want to remove repeated elements, not all duplicates (e.g., `[2, 4, 4, 4, 6, 2]` should turn into `[2, 4, 6, 2]`, not `[2, 4, 6]`. Now it behaves in the way we need.